src/manifest.rs

```diff
@@ -65,262 +65,153 @@ impl From<bare_version::Error> for ManifestParseError {
     }
 }
 
-impl TryFrom<Document> for CargoManifest {
+impl TryFrom<Metadata> for CargoManifest {
     type Error = ManifestParseError;
 
-    fn try_from(map: Document) -> Result<Self, Self::Error> {
-        let minimum_rust_version = minimum_rust_version(&map)?;
+    fn try_from(metadata: Metadata) -> Result<Self, Self::Error> {
+        let minimum_rust_version = find_minimum_rust_version(&metadata)?;
 
         Ok(Self {
             minimum_rust_version,
         })
     }
 }
 
-fn minimum_rust_version(value: &Document) -> Result<Option<BareVersion>, bare_version::Error> {
-    let version = match find_minimum_rust_version(value) {
-        Some(version) => version,
-        None => return Ok(None),
-    };
-
-    Ok(Some(version.parse()?))
-}
-
-/// Parse the minimum supported Rust version (MSRV) from `Cargo.toml` manifest data.
-fn find_minimum_rust_version(document: &Document) -> Option<&str> {
+/// Parse the minimum supported Rust version (MSRV) from `Cargo.toml` metadata.
+fn find_minimum_rust_version(
+    metadata: &Metadata,
+) -> Result<Option<BareVersion>, bare_version::Error> {
     /// Parses the `MSRV` as supported by Cargo since Rust 1.56.0
     ///
     /// [`Cargo`]: https://doc.rust-lang.org/cargo/reference/manifest.html#the-rust-version-field
-    fn find_rust_version(document: &Document) -> Option<&str> {
-        document
-            .as_table()
-            .get("package")
-            .and_then(Item::as_table)
-            .and_then(|package| package.get("rust-version"))
-            .and_then(Item::as_str)
+    fn find_rust_version(metadata: &Metadata) -> Option<&semver::Version> {
+        metadata.root_package()?.rust_version.as_ref()
     }
 
     /// Parses the MSRV as supported by `cargo-msrv`, since prior to the release of Rust
     /// 1.56.0
-    fn find_metadata_msrv(document: &Document) -> Option<&str> {
-        document
-            .as_table()
-            .get("package")
-            .and_then(Item::as_table)
-            .and_then(|package| package.get("metadata"))
-            .and_then(Item::as_table_like)
-            .and_then(|metadata| metadata.get("msrv"))
-            .and_then(Item::as_str)
+    fn find_metadata_msrv(metadata: &Metadata) -> Option<&str> {
+        metadata
+            .root_package()?
+            .metadata
+            .as_object()?
+            .get("msrv")?
+            .as_str()
     }
 
     // Parse the MSRV from the `package.rust-version` key if it exists,
     // and try to fallback to our own `package.metadata.msrv` if it doesn't
-    find_rust_version(document).or_else(|| find_metadata_msrv(document))
+    match find_rust_version(metadata) {
+        Some(version) => Ok(Some(BareVersion::from(version))),
+        None => find_metadata_msrv(metadata)
+            .map(BareVersion::try_from)
+            .transpose(),
```

Comment:
Main changes are here, using `cargo_metadata` to parse the necessary data instead of parsing via `Cargo.toml` only